hathor/p2p/sync_v2/manager.py

```diff
@@ -0,0 +1,1207 @@
+        heights = json.loads(payload)
+        data = []
+        for h in heights:
+            block = self.tx_storage.indexes.height.get(h)
+            if block is None:
+                break
+            data.append((h, block.hex()))
+        payload = json.dumps(data)
+        self.send_message(ProtocolMessages.PEER_BLOCK_HASHES, payload)
+
+    def handle_peer_block_hashes(self, payload: str) -> None:
+        data = json.loads(payload)
+        data = [(h, bytes.fromhex(block_hash)) for (h, block_hash) in data]
+        key = 'peer-block-hashes'
+        deferred = self.deferred_by_key.pop(key, None)
+        if deferred:
+            deferred.callback(data)
+
+    def send_get_next_blocks(self, start_hash: bytes, end_hash: bytes) -> None:
+        payload = json.dumps(dict(
+            start_hash=start_hash.hex(),
+            end_hash=end_hash.hex(),
+        ))
+        self.send_message(ProtocolMessages.GET_NEXT_BLOCKS, payload)
+        self.receiving_stream = True
+
+    def handle_get_next_blocks(self, payload: str) -> None:
+        self.log.debug('handle GET-NEXT-BLOCKS')
+        if self._is_streaming:
+            self.protocol.send_error_and_close_connection('GET-NEXT-BLOCKS received before previous one finished')
+            return
+        data = json.loads(payload)
+        self.send_next_blocks(
+            start_hash=bytes.fromhex(data['start_hash']),
+            end_hash=bytes.fromhex(data['end_hash']),
+        )
+
+    def send_next_blocks(self, start_hash: bytes, end_hash: bytes) -> None:
+        self.log.debug('start GET-NEXT-BLOCKS stream response')
+        # XXX If I don't have this block it will raise TransactionDoesNotExist error. Should I handle this?
+        blk = self.tx_storage.get_transaction(start_hash)
+        assert isinstance(blk, Block)
+        self.blockchain_streaming = BlockchainStreaming(self, blk, end_hash)
+        self.blockchain_streaming.start()
+
+    def send_get_prev_blocks(self, start_hash: bytes, end_hash: bytes) -> None:
+        payload = json.dumps(dict(
+            start_hash=start_hash.hex(),
+            end_hash=end_hash.hex(),
+        ))
+        self.send_message(ProtocolMessages.GET_PREV_BLOCKS, payload)
+        self.receiving_stream = True
+
+    def handle_get_prev_blocks(self, payload: str) -> None:
+        self.log.debug('handle GET-PREV-BLOCKS')
+        if self._is_streaming:
+            self.protocol.send_error_and_close_connection('GET-PREV-BLOCKS received before previous one finished')
+            return
+        data = json.loads(payload)
+        self.send_prev_blocks(
+            start_hash=bytes.fromhex(data['start_hash']),
+            end_hash=bytes.fromhex(data['end_hash']),
+        )
+
+    def send_prev_blocks(self, start_hash: bytes, end_hash: bytes) -> None:
+        self.log.debug('start GET-PREV-BLOCKS stream response')
+        # XXX If I don't have this block it will raise TransactionDoesNotExist error. Should I handle this?
```

Comment:
Same here.

Reply:
Done.

Reply:
I can't see its implementation. Are you sure?

Reply:
This was moved to the sync-checkpoint branch.